### PR TITLE
BDN-780: Auction Controller Crashes Fixes.

### PR DESCRIPTION
**Pull request by zsysenko**

### CODE REVIEW

File: Bidon/Modules/Auction/Controller/ConcurrentAuctionController.swift [hunk 1]
```diff
@@ -59,123 +54,175 @@ final class ConcurrentAuctionController<AdTypeContextType: AdTypeContext>: Aucti
...
     
-    func addOperation(_ operation: any AuctionOperationRequestDemand) {
-        // if we reached timeout we need clear future operations and run finish auction operation
-        if auctionTimeoutReached, let finishAuctionOperation {
-            pendingOperations = []
-            queue.addOperation(finishAuctionOperation)
-            timeoutQueue.cancelAllOperations()
+    deinit {
+        print("deinit swag")
+    }
```
Comment (EvgeniaGorbacheva) [inline Bidon/Modules/Auction/Controller/ConcurrentAuctionController.swift]: Remove test code

File: Bidon/Modules/Auction/Controller/ConcurrentAuctionController.swift [hunk 2]
```diff
@@ -59,123 +54,175 @@ final class ConcurrentAuctionController<AdTypeContextType: AdTypeContext>: Aucti
...
-        let timeoutOperationHandler = BlockOperation { [weak self] in
-            guard let self else { return }
-            self.pendingOperations
-                .compactMap { self.adUnit(from: $0) }
-                .forEach { self.auctionObserver.log(AuctionTimeoutEvent(adUnit: $0)) }
-
-            self.pendingOperations = []
-            self.auctionTimeoutReached = true
+        if adUnit.pricefloor < maxPrice {
+            handlePriceFloorBelowMax()
```
Comment (EvgeniaGorbacheva) [inline Bidon/Modules/Auction/Controller/ConcurrentAuctionController.swift]: handlePriceFloorBelowMax -> handleBelowPricefloor

File: Bidon/Modules/Auction/Controller/ConcurrentAuctionController.swift [hunk 3]
```diff
@@ -59,123 +54,175 @@ final class ConcurrentAuctionController<AdTypeContextType: AdTypeContext>: Aucti
...
+        pendingOperations = []
+        queue.cancelAllOperations()
         
-        // run first ad unit loading
-        let firstOperation = pendingOperations.removeFirst()
-        addOperation(firstOperation)
+        guard let finishAuctionOperation,
+                !finishAuctionOperation.isExecuting,
+                !finishAuctionOperation.isFinished,
+                !finishAuctionOperation.isCancelled else {
```
Comment (EvgeniaGorbacheva) [inline Bidon/Modules/Auction/Controller/ConcurrentAuctionController.swift]: Please put else to the next line

File: Bidon/Modules/Auction/Controller/ConcurrentAuctionController.swift [hunk 4]
```diff
@@ -59,123 +54,175 @@ final class ConcurrentAuctionController<AdTypeContextType: AdTypeContext>: Aucti
...
+            print("finish_auction : Cant finish auction. Finish is already in progress or completed.")
+            return
+        }
+        queue.addOperation(finishAuctionOperation)
+    }
+    
+    private func handlePriceFloorBelowMax() {
+        pendingOperations
+            .compactMap { adUnit(from: $0) }
+            .forEach { logPriceFloorBelowMax(adUnit: $0) }
```
Comment (EvgeniaGorbacheva) [inline Bidon/Modules/Auction/Controller/ConcurrentAuctionController.swift]: This code is called from addOperation method. In this method you check if adunit pricefloor is below maxPrice, and if yes, you call logPriceFloorBelowMax for ALL further ad units

File: Bidon/Modules/Auction/Demands/DemandsTokensManager.swift [hunk 5]
```diff
@@ -44,13 +44,20 @@ final class DemandsTokensManager<AdTypeContextType: AdTypeContext> {
...
+        initializationParameters: AdaptersInitialisationParameters,
+        completion: @escaping ((Result<[BiddingDemandToken], Error>) -> Void)
+    ) {
         var operations = [Operation]()
         for demandId in demands {
             guard
-                let adapter = adapters.first(where: { $0.demandId == demandId && $0.provider is any GenericBiddingDemandProvider }),
+                let adapter = adapters.first(
+                    where: { $0.demandId == demandId && $0.provider is any GenericBiddingDemandProvider }
+                ),
```
Comment (EvgeniaGorbacheva) [inline Bidon/Modules/Auction/Demands/DemandsTokensManager.swift]: Please do not fix codestyle

File: Bidon/Modules/Auction/Demands/DemandsTokensManager.swift [hunk 6]
```diff
@@ -44,13 +44,20 @@ final class DemandsTokensManager<AdTypeContextType: AdTypeContext> {
...
                 let provider = adapter.provider as? any GenericBiddingDemandProvider,
-                let parameters = initializationParameters.adapters.first(where: { $0.demandId == adapter.demandId })
+                let parameters = initializationParameters.adapters.first(
+                    where: { $0.demandId == adapter.demandId }
```
Comment (EvgeniaGorbacheva) [inline Bidon/Modules/Auction/Demands/DemandsTokensManager.swift]: Please do not fix codestyle

File: Bidon/Modules/Auction/Operations/AuctionOperationFinish.swift [hunk 7]
```diff
@@ -55,10 +54,13 @@ where BidType.ProviderType == AdTypeContextType.DemandProviderType, BidType.Dema
         observer.log(CancelAuctionEvent())
         
         findWinner()
+        
         let result = Result<BidType, SdkError>.failure(.cancelled)
         let completion = self.completion
         
-        completion(result)
+        DispatchQueue.main.async {
+            completion(result)
```
Comment (EvgeniaGorbacheva) [inline Bidon/Modules/Auction/Operations/AuctionOperationFinish.swift]: If you integrate Bidon to Appodeal and Appodeal will call cancel() for Bidon's auction, /stats request will not be sent. I told you about this before starting the task, all code in cancel logic must be synchronous

File: Bidon/SDK/Fullscreen/RewardedAd.swift [hunk 8]
```diff
@@ -59,6 +59,11 @@ public final class RewardedAd: NSObject, RewardedAdObject {
         manager.loadAd(pricefloor: pricefloor, auctionKey: auctionKey)
     }
     
+    @objc
+    public func cancel() {
+        manager.cancel()
+    }
```
Comment (EvgeniaGorbacheva) [inline Bidon/SDK/Fullscreen/RewardedAd.swift]: You missed cancel logic for banners

File: Sandbox/Sandbox/Shared/Advertising/Raw/RawAdService.swift [hunk 9]
```diff
@@ -77,6 +77,17 @@ final class RawAdService: NSObject, AdService {
...
     }
     
+    func cancel(adType: AdType) throws {
+        switch adType {
+        case .interstitial:
+            interstitial.cancel()
+        case .rewardedAd:
+            rewardedAd.cancel()
+        default:
+            throw RawAdServiceError.unsupported
```
Comment (EvgeniaGorbacheva) [inline Sandbox/Sandbox/Shared/Advertising/Raw/RawAdService.swift]: You missed banner logic here

File: Bidon/Modules/Auction/Controller/ConcurrentAuctionController.swift (hunk 4 above)
Comment (EvgeniaGorbacheva) [reply Bidon/Modules/Auction/Controller/ConcurrentAuctionController.swift]: Also, in addOperation method you cast adUnit to operation and here you cast it back

File: Bidon/SDK/Fullscreen/FullscreenAdManager.swift [hunk 10]
```diff
@@ -318,6 +318,15 @@ AdaptersFetcherType: AdaptersFetcher<AdTypeContextType> {
         }
     }
     
+    func cancel() {
+        switch state {
+        case .auction(let controller):
+            controller.cancel()
```
Comment (EvgeniaGorbacheva) [inline Bidon/SDK/Fullscreen/FullscreenAdManager.swift]: If I call cancel() when, for example, tokens are being collect, but auction has not been already started, it will be started anyway

File: Bidon/Modules/Auction/Controller/ConcurrentAuctionController.swift [hunk 11]
```diff
@@ -59,123 +54,175 @@ final class ConcurrentAuctionController<AdTypeContextType: AdTypeContext>: Aucti
...
         
-        // run first ad unit loading
-        let firstOperation = pendingOperations.removeFirst()
-        addOperation(firstOperation)
+        guard let finishAuctionOperation,
+                !finishAuctionOperation.isExecuting,
+                !finishAuctionOperation.isFinished,
+                !finishAuctionOperation.isCancelled else {
+            
+            print("finish_auction : Cant finish auction. Finish is already in progress or completed.")
```
Comment (EvgeniaGorbacheva) [inline Bidon/Modules/Auction/Controller/ConcurrentAuctionController.swift]: logger?

File: Bidon/Modules/Auction/Operations/Demand/AuctionOperationRequestBiddingDemand.swift [hunk 12]
```diff
@@ -63,9 +61,12 @@ final class AuctionOperationRequestBiddingDemand<AdTypeContextType: AdTypeContex
...
             adUnitExtrasDecoder: adUnit.extras
         ) { [weak self] result in
-            guard let self = self else { return }
-            defer {
-                self.finish()
+            guard let self else { return }
+            defer { self.finish() }
+            
+            guard !isCancelled else {
+                print("Operation is cancelled due Timeout or Cancel event.")
```
Comment (EvgeniaGorbacheva) [inline Bidon/Modules/Auction/Operations/Demand/AuctionOperationRequestBiddingDemand.swift]: logger?

File: Bidon/Modules/Auction/Operations/Demand/AuctionOperationRequestDirectDemand.swift [hunk 13]
```diff
@@ -36,8 +36,6 @@ final class AuctionOperationRequestDirectDemand<AdTypeContextType: AdTypeContext
     override func main() {
         super.main()
         
-        guard isExecuting else { return }
```
Comment (EvgeniaGorbacheva) [inline Bidon/Modules/Auction/Operations/Demand/AuctionOperationRequestDirectDemand.swift:39]: Why did you remove this code?

File: Bidon/SDK/Fullscreen/FullscreenAdManager.swift [hunk 14]
```diff
@@ -318,6 +318,15 @@ AdaptersFetcherType: AdaptersFetcher<AdTypeContextType> {
...
+        default: break
+            // print("Cant cancel auction. It is not in progress")
```
Comment (EvgeniaGorbacheva) [inline Bidon/SDK/Fullscreen/FullscreenAdManager.swift]: Unused code?

File: Sandbox/Sandbox/FullscreenAd/InterstitialSection.swift [hunk 15]
```diff
@@ -69,6 +69,18 @@ struct InterstitialSection: View {
                 "Events",
                 destination: AdEventsList(events: vm.events)
             )
+            
+            Button(action: cancelAuction) {
+                HStack {
+                    switch vm.state {
+                    case .ready:
+                        Text("Revoke")
+                            .disabled(false)
```
Comment (EvgeniaGorbacheva) [inline Sandbox/Sandbox/FullscreenAd/InterstitialSection.swift]: Disabled text?

File: Bidon/Modules/Auction/Controller/ConcurrentAuctionController.swift (hunk 4 above)
Comment (zsysenko) [reply Bidon/Modules/Auction/Controller/ConcurrentAuctionController.swift]: > > This code is called from addOperation method. In this method you check if adunit pricefloor is below maxPrice, and if yes, you call logPriceFloorBelowMax for ALL further ad units

Yes, correct. What is the problem here?

> Also, in addOperation method you cast adUnit to operation and here you cast it back

Didnt get the point. What you mean here?

File: Bidon/Modules/Auction/Operations/AuctionOperationFinish.swift (hunk 7 above)
Comment (zsysenko) [reply Bidon/Modules/Auction/Operations/AuctionOperationFinish.swift]: Didnt get the point here. Why it works like that?

File: Bidon/Modules/Auction/Operations/Demand/AuctionOperationRequestDirectDemand.swift (hunk 13 above)
Comment (zsysenko) [reply Bidon/Modules/Auction/Operations/Demand/AuctionOperationRequestDirectDemand.swift:39]: Looks like it does nothing. For what it should be here?

File: Bidon/Modules/Auction/Demands/DemandsTokensManager.swift (hunk 5 above)
Comment (zsysenko) [reply Bidon/Modules/Auction/Demands/DemandsTokensManager.swift]: Should I bring it back?

File: Bidon/Modules/Auction/Controller/ConcurrentAuctionController.swift (hunk 2 above)
Comment (zsysenko) [reply Bidon/Modules/Auction/Controller/ConcurrentAuctionController.swift]: What you mean by that rename?

File: Bidon/Modules/Auction/Controller/ConcurrentAuctionController.swift (hunk 4 above)
Comment (EvgeniaGorbacheva) [reply Bidon/Modules/Auction/Controller/ConcurrentAuctionController.swift]: >What is the problem here?

You may have adunits array not sorted by ecpm, so there can be a case, when, for example, maxPrice = 5, pending ad units pricefloors can be [4, 3, 1, 20], but looks like you will set for all of them status BELOW_PRICEFLOOR

>Didnt get the point

This method is called only from addOperation method. Inside addOperation you cast adUnit to Operation and then here in handlePriceFloorBelowMax you cast operation back to adUnit

File: Bidon/Modules/Auction/Demands/DemandsTokensManager.swift (hunk 5 above)
Comment (EvgeniaGorbacheva) [reply Bidon/Modules/Auction/Demands/DemandsTokensManager.swift]: Yes please

File: Bidon/Modules/Auction/Controller/ConcurrentAuctionController.swift (hunk 4 above)
Comment (zsysenko) [reply Bidon/Modules/Auction/Controller/ConcurrentAuctionController.swift]: > You may have adunits array not sorted by ecpm, so there can be a case, when, for example, maxPrice = 5, pending ad units pricefloors can be [4, 3, 1, 20], but looks like you will set for all of them status BELOW_PRICEFLOOR

Thought it was order by pricefloor. Will fix it

File: Bidon/Modules/Auction/Operations/AuctionOperationFinish.swift (hunk 7 above)
Comment (EvgeniaGorbacheva) [reply Bidon/Modules/Auction/Operations/AuctionOperationFinish.swift]: For example: Appodeal Bidon adapter contains reference to Interstitial. When we call notifyLoss (aka cancel) in Appodeal adapter, we need to cancel the auction and then send /stats request. But if your code will be async, at the moment you call DispatchQueue.main.async the Interstitial object will released and result will not be handled. And in FullscreenAdManager when the closure will be called guard let self checking will fail, because Interstitial object (and FullscreenAdManager) will be both released. Just remove DispatchQueue.main.async and all will be ok

File: Bidon/Modules/Auction/Operations/Demand/AuctionOperationRequestDirectDemand.swift (hunk 13 above)
Comment (EvgeniaGorbacheva) [reply Bidon/Modules/Auction/Operations/Demand/AuctionOperationRequestDirectDemand.swift:39]: Maybe for previous versions it was added to avoid smth but looks like you are right

File: Bidon/Modules/Auction/Controller/ConcurrentAuctionController.swift (hunk 2 above)
Comment (EvgeniaGorbacheva) [reply Bidon/Modules/Auction/Controller/ConcurrentAuctionController.swift]: Just matter of taste, below max sounds like we have **predefined** maximal pricefloor and we check every ad unit to match it :)

File: Bidon/Modules/Auction/Operations/AuctionOperationFinish.swift (hunk 7 above)
Comment (zsysenko) [reply Bidon/Modules/Auction/Operations/AuctionOperationFinish.swift]: Brought it back.

**Review by EvgeniaGorbacheva — APPROVED** [review]
